File: LineUp.java
import java.util.*;
import java.io.*;

public class LineUp{
   public static void main(String[] args) 
   {
      Scanner sc = new Scanner(System.in);
      try{
         sc = new Scanner(new File("lineup.in"));
      }
      catch(FileNotFoundException e){
         System.out.println(e);
         
      }
      int num = sc.nextInt();
      ArrayList<String> a = new ArrayList<>();
      for(int i = 0; i<num+1; i++){
         a.add(sc.nextLine());
      }
      a.remove(0);
      System.out.println(a);
      String[] names = {"Bessie", "Buttercup", "Belinda", "Beatrice", "Bella", "Blue", "Betsy","Sue"};
      ArrayList<Pair> pairs = new ArrayList<>();
   
      for(String s:a){
         String[] strA = s.split(" must be milked beside ");
         pairs.add(new Pair(strA[0], strA[1]));
      }
     //  BSTobject<String> tree = null;
   //       tree.add(pairs.get(0).getName1());
   //       tree.add(tree.contains(pairs.get(0).getName1()), pairs.get(0).getName2());
   //   
      String[] sa = checkAllRecursive(names.length, names, pairs);
      try{
         FileWriter myWriter = new FileWriter("lineup.out");
         for(String s: sa)
            myWriter.write(s+"\n");
         myWriter.close();
      }
      catch(IOException e){
      }
   
    
   }
   public static boolean works(String[] line, ArrayList<Pair> pa){
      int num = 0;
   
      for(int i = 1; i<line.length-1; i++){
         for(Pair p : pa){
            if(i<line.length-1&&((line[i].equals(p.getName1())&&line[i-1].equals(p.getName2())) ||(line[i].equals(p.getName2())&&line[i-1].equals(p.getName1()))))
               num++;
            if(i<line.length-1&&((line[i].equals(p.getName1())&&line[i+1].equals(p.getName2()))||(line[i].equals(p.getName2())&&line[i+1].equals(p.getName1()))))
            {
               num++;
               i+=2;
            }
         }
         if(num==pa.size())
            return true;
      }
      return false;
   }
   public static String[]  checkAllRecursive(  int n, String[] elements, ArrayList<Pair> pa) {
      ArrayList<String[]> array = new ArrayList<>();
      int[] indexes = new int[n];
      for (int i = 0; i < n; i++) {
         indexes[i] = 0;
      }
      int i = 0;
      while (i < n) {
         if(works(elements, pa)){
            if(!array.contains(elements))
               array.add(elements);
            
            }
            
         if (indexes[i] < i) {
            elements = swap(elements, i % 2 == 0 ?  0: indexes[i], i);
            indexes[i]++;
            i = 0;
         }
         else {
            indexes[i] = 0;
            i++;
         }
      }
     String first = array.get(0)[0];
         int index = 0;
      for( i = 1; i<array.size()-1; i++){
         int x = 0;
         if(array.get(i)[x].compareTo(first)>0){
            first = array.get(i)[x];
            index = i;
            }
         else if(array.get(i)[x].compareTo(first)==0){
            while(x<7){
               x++;
               if(array.get(i)[x].compareTo(first)>0){
                  first = array.get(i)[x];
                   index = i;}
               else if(array.get(i)[x].compareTo(first)==0)
                  x++;
              }
         }
      }
      return  array.get(index);
   
   }
   public static String[] swap(String[] input, int a, int b) {
      String tmp = input[a];
      input[a] = input[b];
      input[b] = tmp;
      return input;
   }
  
}
class Pair{
   String name1, name2;
   Pair(String n1, String n2){
      this.name1 = n1;
      this.name2 = n2;
   }
   String getName1(){
      return name1;
   }
   String getName2(){
      return name2;
   }
}

File: lineup.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <set>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

int alphaScore(const vector<vector<string>>& input) {
    vector<string> s;
    int score = 0;
    for (const vector<string>& vec : input)
        for (const string& cow : vec) s.emplace_back(cow);
    for (int i = 0; i < s.size(); i++)
        for (int j = i + 1; j < s.size(); j++) {
            if (s[i] > s[j]) score++;
        }
    return score;
}

void insertList(vector<vector<string>>& output, const vector<string>& insert) {
    vector<vector<string>> temp(output);
    temp.insert(temp.begin(), insert);
    int min_score = alphaScore(temp);
    int score;
    vector<vector<string>> mostAlpha(temp);
    for (int i = 1; i <= output.size(); i++) {
        temp = output;
        temp.insert(temp.begin() + i, insert);
        score = alphaScore(temp);
        if (score < min_score) {
            min_score = score;
            mostAlpha = temp;
        }
    }

    output = mostAlpha;
    // for (auto iter = output.begin(); iter != output.end(); ++iter) {
    //     if ((*iter)[0] > *insert.begin()) {
    //         temp.insert(iter, insert);
    //         return;
    //     }
    // }
    // output.emplace_back(insert);
}

int main() {
    vector<string> cows{"Beatrice",   //
                        "Belinda",    //
                        "Bella",      //
                        "Bessie",     //
                        "Betsy",      //
                        "Blue",       //
                        "Buttercup",  //
                        "Sue"};

    ifstream file("lineup.in", ios::in);
    int len;
    file >> len;
    string first, must, be, milked, beside, second;

    unordered_map<string, set<string>> adj;

    for (int i = 0; i < len; i++) {
        file >> first >> must >> be >> milked >> beside >> second;
        cows.erase(remove(cows.begin(), cows.end(), first), cows.end());
        cows.erase(remove(cows.begin(), cows.end(), second), cows.end());
        adj[second].insert(first);
    }

    vector<vector<string>> output;
    for (const string& cow : cows) {
        output.emplace_back(vector<string>{cow});
    }
    for (auto const& s : adj) {
        vector<string> insert;
        if (s.second.size() > 1) {
            auto second = s.second.begin();
            advance(second, 1);
            if (*s.second.begin() < *second) {
                insert.emplace_back(*s.second.begin());
                insert.emplace_back(s.first);
                insert.emplace_back(*second);
            } else {
                insert.emplace_back(*second);
                insert.emplace_back(s.first);
                insert.emplace_back(*s.second.begin());
            }
        } else {
            if (s.first < *s.second.begin()) {
                insert.emplace_back(s.first);
                insert.emplace_back(*s.second.begin());
            } else {
                insert.emplace_back(*s.second.begin());
                insert.emplace_back(s.first);
            }
        }
        insertList(output, insert);
    }

    ofstream out("lineup.out", ios::out);

    for (const vector<string>& cowvec : output) {
        for (const string& cow : cowvec) out << cow << '\n';
    }

    return 0;
}